File: leet-code/convertBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var convertBST = function (root) {
  if (root == null) {
    return null
  }
  var sum = 0
  function sub (root) {
    if (root == null) {
      return null
    }
    sub(root.right)
    sum += root.val
    root.val = sum
    sub(root.left)
    return root
  }
  return sub(root)
}

var tree = {
  val: 2,
  left: { val: 1, left: null, right: null },
  right: { val: 3, left: null, right: null }
}

console.log(convertBST(tree))

File: leet-code/convertBST.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func convertBST(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	var sum int
	var f func(root *TreeNode) *TreeNode
	f = func(root *TreeNode) *TreeNode {
		if root == nil {
			return nil
		}
		f(root.Right)
		sum += root.Val
		root.Val = sum
		f(root.Left)
		return root
	}
	return f(root)
}
